File: examples/java/TopologicalSortExamples.java
import java.util.*;

public class TopologicalSortExamples {
    
    // 1. Course Schedule
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) graph.add(new ArrayList<>());
        
        for (int[] pre : prerequisites) {
            graph.get(pre[1]).add(pre[0]);
            indegree[pre[0]]++;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }
        
        int count = 0;
        while (!queue.isEmpty()) {
            int course = queue.poll();
            count++;
            for (int next : graph.get(course)) {
                if (--indegree[next] == 0) queue.offer(next);
            }
        }
        return count == numCourses;
    }
    
    // 2. Course Schedule II
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) graph.add(new ArrayList<>());
        
        for (int[] pre : prerequisites) {
            graph.get(pre[1]).add(pre[0]);
            indegree[pre[0]]++;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }
        
        int[] result = new int[numCourses];
        int idx = 0;
        while (!queue.isEmpty()) {
            int course = queue.poll();
            result[idx++] = course;
            for (int next : graph.get(course)) {
                if (--indegree[next] == 0) queue.offer(next);
            }
        }
        return idx == numCourses ? result : new int[0];
    }
    
    // 3. Alien Dictionary
    public String alienOrder(String[] words) {
        Map<Character, Set<Character>> graph = new HashMap<>();
        Map<Character, Integer> indegree = new HashMap<>();
        
        for (String word : words) {
            for (char c : word.toCharArray()) {
                indegree.put(c, 0);
                graph.put(c, new HashSet<>());
            }
        }
        
        for (int i = 0; i < words.length - 1; i++) {
            String w1 = words[i], w2 = words[i + 1];
            if (w1.length() > w2.length() && w1.startsWith(w2)) return "";
            
            for (int j = 0; j < Math.min(w1.length(), w2.length()); j++) {
                char c1 = w1.charAt(j), c2 = w2.charAt(j);
                if (c1 != c2) {
                    if (!graph.get(c1).contains(c2)) {
                        graph.get(c1).add(c2);
                        indegree.put(c2, indegree.get(c2) + 1);
                    }
                    break;
                }
            }
        }
        
        Queue<Character> queue = new LinkedList<>();
        for (char c : indegree.keySet()) {
            if (indegree.get(c) == 0) queue.offer(c);
        }
        
        StringBuilder result = new StringBuilder();
        while (!queue.isEmpty()) {
            char c = queue.poll();
            result.append(c);
            for (char next : graph.get(c)) {
                indegree.put(next, indegree.get(next) - 1);
                if (indegree.get(next) == 0) queue.offer(next);
            }
        }
        
        return result.length() == indegree.size() ? result.toString() : "";
    }
    
    // 4. Minimum Height Trees
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (n == 1) return Arrays.asList(0);
        
        List<Set<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new HashSet<>());
        
        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }
        
        Queue<Integer> leaves = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (graph.get(i).size() == 1) leaves.offer(i);
        }
        
        int remaining = n;
        while (remaining > 2) {
            int leafCount = leaves.size();
            remaining -= leafCount;
            for (int i = 0; i < leafCount; i++) {
                int leaf = leaves.poll();
                int neighbor = graph.get(leaf).iterator().next();
                graph.get(neighbor).remove(leaf);
                if (graph.get(neighbor).size() == 1) leaves.offer(neighbor);
            }
        }
        
        return new ArrayList<>(leaves);
    }
    
    // 5. Sequence Reconstruction
    public boolean sequenceReconstruction(int[] nums, List<List<Integer>> sequences) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        Map<Integer, Integer> indegree = new HashMap<>();
        
        for (int num : nums) {
            graph.put(num, new HashSet<>());
            indegree.put(num, 0);
        }
        
        for (List<Integer> seq : sequences) {
            for (int i = 0; i < seq.size() - 1; i++) {
                int from = seq.get(i), to = seq.get(i + 1);
                if (!graph.containsKey(from) || !graph.containsKey(to)) return false;
                if (!graph.get(from).contains(to)) {
                    graph.get(from).add(to);
                    indegree.put(to, indegree.get(to) + 1);
                }
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int num : indegree.keySet()) {
            if (indegree.get(num) == 0) queue.offer(num);
        }
        
        int idx = 0;
        while (!queue.isEmpty()) {
            if (queue.size() > 1) return false;
            int num = queue.poll();
            if (nums[idx++] != num) return false;
            for (int next : graph.get(num)) {
                indegree.put(next, indegree.get(next) - 1);
                if (indegree.get(next) == 0) queue.offer(next);
            }
        }
        
        return idx == nums.length;
    }
    
    // 6. All Ancestors of a Node in DAG
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[n];
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        
        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            indegree[edge[1]]++;
        }
        
        List<Set<Integer>> ancestors = new ArrayList<>();
        for (int i = 0; i < n; i++) ancestors.add(new HashSet<>());
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }
        
        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int child : graph.get(node)) {
                ancestors.get(child).add(node);
                ancestors.get(child).addAll(ancestors.get(node));
                if (--indegree[child] == 0) queue.offer(child);
            }
        }
        
        List<List<Integer>> result = new ArrayList<>();
        for (Set<Integer> set : ancestors) {
            List<Integer> list = new ArrayList<>(set);
            Collections.sort(list);
            result.add(list);
        }
        return result;
    }
    
    // 7. Parallel Courses
    public int minimumSemesters(int n, int[][] relations) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[n + 1];
        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());
        
        for (int[] relation : relations) {
            graph.get(relation[0]).add(relation[1]);
            indegree[relation[1]]++;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 1; i <= n; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }
        
        int semesters = 0, studied = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            semesters++;
            for (int i = 0; i < size; i++) {
                int course = queue.poll();
                studied++;
                for (int next : graph.get(course)) {
                    if (--indegree[next] == 0) queue.offer(next);
                }
            }
        }
        
        return studied == n ? semesters : -1;
    }
    
    // 8. Sort Items by Groups Respecting Dependencies
    public int[] sortItems(int n, int m, int[] group, List<List<Integer>> beforeItems) {
        int groupId = m;
        for (int i = 0; i < n; i++) {
            if (group[i] == -1) group[i] = groupId++;
        }
        
        List<List<Integer>> itemGraph = new ArrayList<>();
        List<List<Integer>> groupGraph = new ArrayList<>();
        int[] itemIndegree = new int[n];
        int[] groupIndegree = new int[groupId];
        
        for (int i = 0; i < n; i++) itemGraph.add(new ArrayList<>());
        for (int i = 0; i < groupId; i++) groupGraph.add(new ArrayList<>());
        
        for (int i = 0; i < n; i++) {
            for (int prev : beforeItems.get(i)) {
                itemGraph.get(prev).add(i);
                itemIndegree[i]++;
                
                if (group[prev] != group[i]) {
                    groupGraph.get(group[prev]).add(group[i]);
                    groupIndegree[group[i]]++;
                }
            }
        }
        
        List<Integer> itemOrder = topologicalSort(itemGraph, itemIndegree);
        List<Integer> groupOrder = topologicalSort(groupGraph, groupIndegree);
        
        if (itemOrder.isEmpty() || groupOrder.isEmpty()) return new int[0];
        
        Map<Integer, List<Integer>> groupToItems = new HashMap<>();
        for (int item : itemOrder) {
            groupToItems.computeIfAbsent(group[item], k -> new ArrayList<>()).add(item);
        }
        
        List<Integer> result = new ArrayList<>();
        for (int g : groupOrder) {
            result.addAll(groupToItems.getOrDefault(g, new ArrayList<>()));
        }
        
        return result.stream().mapToInt(i -> i).toArray();
    }
    
    private List<Integer> topologicalSort(List<List<Integer>> graph, int[] indegree) {
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }
        
        List<Integer> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            int node = queue.poll();
            result.add(node);
            for (int next : graph.get(node)) {
                if (--indegree[next] == 0) queue.offer(next);
            }
        }
        
        return result.size() == graph.size() ? result : new ArrayList<>();
    }
    
    // 9. Build a Matrix With Conditions
    public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
        List<Integer> rowOrder = topologicalSort(k, rowConditions);
        List<Integer> colOrder = topologicalSort(k, colConditions);
        
        if (rowOrder.isEmpty() || colOrder.isEmpty()) return new int[0][0];
        
        Map<Integer, Integer> rowPos = new HashMap<>();
        Map<Integer, Integer> colPos = new HashMap<>();
        
        for (int i = 0; i < k; i++) {
            rowPos.put(rowOrder.get(i), i);
            colPos.put(colOrder.get(i), i);
        }
        
        int[][] result = new int[k][k];
        for (int i = 1; i <= k; i++) {
            result[rowPos.get(i)][colPos.get(i)] = i;
        }
        
        return result;
    }
    
    private List<Integer> topologicalSort(int k, int[][] conditions) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[k + 1];
        for (int i = 0; i <= k; i++) graph.add(new ArrayList<>());
        
        for (int[] condition : conditions) {
            graph.get(condition[0]).add(condition[1]);
            indegree[condition[1]]++;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 1; i <= k; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }
        
        List<Integer> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            int node = queue.poll();
            result.add(node);
            for (int next : graph.get(node)) {
                if (--indegree[next] == 0) queue.offer(next);
            }
        }
        
        return result.size() == k ? result : new ArrayList<>();
    }
    
    // 10. Find All Possible Recipes
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        Set<String> supplySet = new HashSet<>(Arrays.asList(supplies));
        Map<String, List<String>> graph = new HashMap<>();
        Map<String, Integer> indegree = new HashMap<>();
        
        for (int i = 0; i < recipes.length; i++) {
            String recipe = recipes[i];
            indegree.put(recipe, 0);
            for (String ingredient : ingredients.get(i)) {
                if (!supplySet.contains(ingredient)) {
                    graph.computeIfAbsent(ingredient, k -> new ArrayList<>()).add(recipe);
                    indegree.put(recipe, indegree.get(recipe) + 1);
                }
            }
        }
        
        Queue<String> queue = new LinkedList<>();
        for (String recipe : indegree.keySet()) {
            if (indegree.get(recipe) == 0) queue.offer(recipe);
        }
        
        List<String> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            String recipe = queue.poll();
            result.add(recipe);
            for (String next : graph.getOrDefault(recipe, new ArrayList<>())) {
                indegree.put(next, indegree.get(next) - 1);
                if (indegree.get(next) == 0) queue.offer(next);
            }
        }
        
        return result;
    }
    
    public static void main(String[] args) {
        TopologicalSortExamples examples = new TopologicalSortExamples();
        
        System.out.println("=== Topological Sort Examples ===");
        
        // Test 1: Course Schedule
        System.out.println("Can Finish(2, [[1,0]]): " + examples.canFinish(2, new int[][]{{1,0}}));
        
        // Test 2: Course Schedule II
        System.out.println("Find Order(2, [[1,0]]): " + Arrays.toString(examples.findOrder(2, new int[][]{{1,0}})));
        
        // Test 3: Alien Dictionary
        System.out.println("Alien Order(['wrt','wrf','er','ett','rftt']): " + 
            examples.alienOrder(new String[]{"wrt","wrf","er","ett","rftt"}));
        
        // Test 4: Minimum Height Trees
        System.out.println("Min Height Trees(4, [[1,0],[1,2],[1,3]]): " + 
            examples.findMinHeightTrees(4, new int[][]{{1,0},{1,2},{1,3}}));
        
        // Test 5: Sequence Reconstruction
        List<List<Integer>> sequences = Arrays.asList(Arrays.asList(1,2), Arrays.asList(1,3));
        System.out.println("Sequence Reconstruction([1,2,3], [[1,2],[1,3]]): " + 
            examples.sequenceReconstruction(new int[]{1,2,3}, sequences));
        
        // Test 6: All Ancestors
        System.out.println("Get Ancestors(8, [[0,3],[0,4],[1,3],[2,4],[2,7],[3,5],[3,6],[3,7],[4,6]]): " + 
            examples.getAncestors(8, new int[][]{{0,3},{0,4},{1,3},{2,4},{2,7},{3,5},{3,6},{3,7},{4,6}}));
        
        // Test 7: Parallel Courses
        System.out.println("Minimum Semesters(3, [[1,3],[2,3]]): " + 
            examples.minimumSemesters(3, new int[][]{{1,3},{2,3}}));
        
        // Test 8: Sort Items (simplified test)
        System.out.println("Sort Items: Complex test case");
        
        // Test 9: Build Matrix
        System.out.println("Build Matrix: Complex test case");
        
        // Test 10: Find All Recipes
        String[] recipes = {"bread"};
        List<List<String>> ingredients = Arrays.asList(Arrays.asList("yeast","flour"));
        String[] supplies = {"yeast","flour","corn"};
        System.out.println("Find All Recipes: " + examples.findAllRecipes(recipes, ingredients, supplies));
    }
}

File: examples/python/topological_sort_examples.py
from collections import defaultdict, deque

# 1. Can Finish
def can_finish(num_courses, prerequisites):
    indegree = [0] * num_courses
    graph = defaultdict(list)
    
    for course, pre in prerequisites:
        graph[pre].append(course)
        indegree[course] += 1
    
    queue = deque([i for i in range(num_courses) if indegree[i] == 0])
    count = 0
    
    while queue:
        course = queue.popleft()
        count += 1
        for next_course in graph[course]:
            indegree[next_course] -= 1
            if indegree[next_course] == 0:
                queue.append(next_course)
    
    return count == num_courses

# 2. Find Order
def find_order(num_courses, prerequisites):
    indegree = [0] * num_courses
    graph = defaultdict(list)
    
    for course, pre in prerequisites:
        graph[pre].append(course)
        indegree[course] += 1
    
    queue = deque([i for i in range(num_courses) if indegree[i] == 0])
    result = []
    
    while queue:
        course = queue.popleft()
        result.append(course)
        for next_course in graph[course]:
            indegree[next_course] -= 1
            if indegree[next_course] == 0:
                queue.append(next_course)
    
    return result if len(result) == num_courses else []

# 3. Alien Order
def alien_order(words):
    graph = defaultdict(set)
    indegree = {c: 0 for word in words for c in word}
    
    for i in range(len(words) - 1):
        w1, w2 = words[i], words[i + 1]
        if len(w1) > len(w2) and w1.startswith(w2):
            return ""
        
        for j in range(min(len(w1), len(w2))):
            c1, c2 = w1[j], w2[j]
            if c1 != c2:
                if c2 not in graph[c1]:
                    graph[c1].add(c2)
                    indegree[c2] += 1
                break
    
    queue = deque([c for c in indegree if indegree[c] == 0])
    result = []
    
    while queue:
        c = queue.popleft()
        result.append(c)
        for next_c in graph[c]:
            indegree[next_c] -= 1
            if indegree[next_c] == 0:
                queue.append(next_c)
    
    return ''.join(result) if len(result) == len(indegree) else ""

# 4. Find Min Height Trees
def find_min_height_trees(n, edges):
    if n == 1:
        return [0]
    
    graph = defaultdict(set)
    for a, b in edges:
        graph[a].add(b)
        graph[b].add(a)
    
    leaves = deque([i for i in range(n) if len(graph[i]) == 1])
    remaining = n
    
    while remaining > 2:
        leaf_count = len(leaves)
        remaining -= leaf_count
        for _ in range(leaf_count):
            leaf = leaves.popleft()
            neighbor = graph[leaf].pop()
            graph[neighbor].remove(leaf)
            if len(graph[neighbor]) == 1:
                leaves.append(neighbor)
    
    return list(leaves)

# 5. Sequence Reconstruction
def sequence_reconstruction(nums, sequences):
    graph = defaultdict(set)
    indegree = {num: 0 for num in nums}
    
    for seq in sequences:
        for i in range(len(seq) - 1):
            from_num, to_num = seq[i], seq[i + 1]
            if from_num not in indegree or to_num not in indegree:
                return False
            if to_num not in graph[from_num]:
                graph[from_num].add(to_num)
                indegree[to_num] += 1
    
    queue = deque([num for num in indegree if indegree[num] == 0])
    idx = 0
    
    while queue:
        if len(queue) > 1:
            return False
        num = queue.popleft()
        if nums[idx] != num:
            return False
        idx += 1
        for next_num in graph[num]:
            indegree[next_num] -= 1
            if indegree[next_num] == 0:
                queue.append(next_num)
    
    return idx == len(nums)

# 6. Get Ancestors
def get_ancestors(n, edges):
    graph = defaultdict(list)
    indegree = [0] * n
    
    for a, b in edges:
        graph[a].append(b)
        indegree[b] += 1
    
    ancestors = [set() for _ in range(n)]
    queue = deque([i for i in range(n) if indegree[i] == 0])
    
    while queue:
        node = queue.popleft()
        for child in graph[node]:
            ancestors[child].add(node)
            ancestors[child].update(ancestors[node])
            indegree[child] -= 1
            if indegree[child] == 0:
                queue.append(child)
    
    return [sorted(list(ancestor_set)) for ancestor_set in ancestors]

# 7. Minimum Semesters
def minimum_semesters(n, relations):
    graph = defaultdict(list)
    indegree = [0] * (n + 1)
    
    for pre, course in relations:
        graph[pre].append(course)
        indegree[course] += 1
    
    queue = deque([i for i in range(1, n + 1) if indegree[i] == 0])
    semesters = 0
    studied = 0
    
    while queue:
        size = len(queue)
        semesters += 1
        for _ in range(size):
            course = queue.popleft()
            studied += 1
            for next_course in graph[course]:
                indegree[next_course] -= 1
                if indegree[next_course] == 0:
                    queue.append(next_course)
    
    return semesters if studied == n else -1

# 8. Sort Items
def sort_items(n, m, group, before_items):
    group_id = m
    for i in range(n):
        if group[i] == -1:
            group[i] = group_id
            group_id += 1
    
    item_graph = defaultdict(list)
    group_graph = defaultdict(list)
    item_indegree = [0] * n
    group_indegree = [0] * group_id
    
    for i in range(n):
        for prev in before_items[i]:
            item_graph[prev].append(i)
            item_indegree[i] += 1
            
            if group[prev] != group[i]:
                group_graph[group[prev]].append(group[i])
                group_indegree[group[i]] += 1
    
    def topological_sort(graph, indegree):
        queue = deque([i for i in range(len(indegree)) if indegree[i] == 0])
        result = []
        while queue:
            node = queue.popleft()
            result.append(node)
            for next_node in graph[node]:
                indegree[next_node] -= 1
                if indegree[next_node] == 0:
                    queue.append(next_node)
        return result if len(result) == len(indegree) else []
    
    item_order = topological_sort(item_graph, item_indegree)
    group_order = topological_sort(group_graph, group_indegree)
    
    if not item_order or not group_order:
        return []
    
    group_to_items = defaultdict(list)
    for item in item_order:
        group_to_items[group[item]].append(item)
    
    result = []
    for g in group_order:
        result.extend(group_to_items[g])
    
    return result

# 9. Build Matrix
def build_matrix(k, row_conditions, col_conditions):
    def topological_sort(conditions):
        graph = defaultdict(list)
        indegree = [0] * (k + 1)
        
        for a, b in conditions:
            graph[a].append(b)
            indegree[b] += 1
        
        queue = deque([i for i in range(1, k + 1) if indegree[i] == 0])
        result = []
        
        while queue:
            node = queue.popleft()
            result.append(node)
            for next_node in graph[node]:
                indegree[next_node] -= 1
                if indegree[next_node] == 0:
                    queue.append(next_node)
        
        return result if len(result) == k else []
    
    row_order = topological_sort(row_conditions)
    col_order = topological_sort(col_conditions)
    
    if not row_order or not col_order:
        return []
    
    row_pos = {num: i for i, num in enumerate(row_order)}
    col_pos = {num: i for i, num in enumerate(col_order)}
    
    result = [[0] * k for _ in range(k)]
    for i in range(1, k + 1):
        result[row_pos[i]][col_pos[i]] = i
    
    return result

# 10. Find All Recipes
def find_all_recipes(recipes, ingredients, supplies):
    supply_set = set(supplies)
    graph = defaultdict(list)
    indegree = defaultdict(int)
    
    for i, recipe in enumerate(recipes):
        for ingredient in ingredients[i]:
            if ingredient not in supply_set:
                graph[ingredient].append(recipe)
                indegree[recipe] += 1
    
    queue = deque([recipe for recipe in recipes if indegree[recipe] == 0])
    result = []
    
    while queue:
        recipe = queue.popleft()
        result.append(recipe)
        for next_recipe in graph[recipe]:
            indegree[next_recipe] -= 1
            if indegree[next_recipe] == 0:
                queue.append(next_recipe)
    
    return result

def run_examples():
    print("=== Topological Sort Examples ===")
    
    # Test 1: Course Schedule
    print(f"Can Finish(2, [[1,0]]): {can_finish(2, [[1,0]])}")
    
    # Test 2: Course Schedule II
    print(f"Find Order(2, [[1,0]]): {find_order(2, [[1,0]])}")
    
    # Test 3: Alien Dictionary
    print(f"Alien Order(['wrt','wrf','er','ett','rftt']): {alien_order(['wrt','wrf','er','ett','rftt'])}")
    
    # Test 4: Minimum Height Trees
    print(f"Min Height Trees(4, [[1,0],[1,2],[1,3]]): {find_min_height_trees(4, [[1,0],[1,2],[1,3]])}")
    
    # Test 5: Sequence Reconstruction
    print(f"Sequence Reconstruction([1,2,3], [[1,2],[1,3]]): {sequence_reconstruction([1,2,3], [[1,2],[1,3]])}")
    
    # Test 6: All Ancestors
    print(f"Get Ancestors(8, [[0,3],[0,4],[1,3],[2,4],[2,7],[3,5],[3,6],[3,7],[4,6]]): {get_ancestors(8, [[0,3],[0,4],[1,3],[2,4],[2,7],[3,5],[3,6],[3,7],[4,6]])}")
    
    # Test 7: Parallel Courses
    print(f"Minimum Semesters(3, [[1,3],[2,3]]): {minimum_semesters(3, [[1,3],[2,3]])}")
    
    # Test 8: Sort Items
    print(f"Sort Items: Complex test case")
    
    # Test 9: Build Matrix
    print(f"Build Matrix: Complex test case")
    
    # Test 10: Find All Recipes
    print(f"Find All Recipes: {find_all_recipes(['bread'], [['yeast','flour']], ['yeast','flour','corn'])}")

if __name__ == "__main__":
    run_examples()